### PR TITLE
Fix ICE on i686 when calling immediate() on OperandValue::Ref in return

### LINKED ISSUE #38727: ICE; two sources of truth on how stuff is returned do not agree with each other
Consider this code:

```rust
#![feature(no_core, lang_items,start)]
#![no_std]
#![no_core]
#[lang="sized"] trait Sized {}
#[lang="copy"] trait Copy {}

#[repr(u64)]
enum A {
    A = 0u64,
    B = !0u64,
}

fn cmp() -> A {
    A::B
}

#[start]
fn main(b:isize, a: *const *const u8)->isize { 0 }
```

if compiled targetting `i686` target, this ICEs with

```
error: internal compiler error: ../src/librustc_trans/mir/operand.rs:82: impossible case reached
```

---

This is because ABI code claims that this u64 enum has to be returned via value (i.e. `ret i64 %smth`):

```
FnType { args: [], ret: ArgType { kind: Direct, original_ty: i64, ty: i64, signedness: Some(false), cast: None, ... }
```

Whereas MIR only has `OperandRef(Ref(...))` because the [`common::is_immediate_type`](https://github.com/rust-lang/rust/blob/6dcf51aae61dcad1dbe7eef3766df2facef73816/src/librustc_trans/mir/operand.rs#L158) returns false for this enum. [Calling `immediate()` on this `OperandRef`](https://github.com/rust-lang/rust/blob/6dcf51aae61dcad1dbe7eef3766df2facef73816/src/librustc_trans/mir/block.rs#L223) then causes this ICE.

This is a regression from the old->MIR trans move.

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
Fixes #38727, and adds a test case.

r? @eddyb

### CODE REVIEW

File: src/librustc_trans/mir/block.rs [hunk 1]
```diff
@@ -223,7 +223,12 @@ impl<'a, 'tcx> MirContext<'a, 'tcx> {
                     load
                 } else {
                     let op = self.trans_consume(&bcx, &mir::Lvalue::Local(mir::RETURN_POINTER));
-                    op.pack_if_pair(&bcx).immediate()
+                    let op_ref = op.pack_if_pair(&bcx);
```
Comment (eddyb) [inline src/librustc_trans/mir/block.rs]: I'd move this into the `else`.

**Comment by nikomatsakis** [comment]
Approved for beta. Small patch, regression.

cc @rust-lang/compiler

**Comment by alexcrichton** [comment]
@bors: p=1 

This is a backport

**Comment by eddyb** [comment]
@bors retry
* cloning LLVM failed

**Comment by alexcrichton** [comment]
@bors: retry

* osx network failure

**Pull request merged**